Feature/attendance api/attendance management

hyeneung opened the pull request:
## #️⃣연관된 이슈

- this closes #13


## 📝작업 내용
- 이전 PR 코드 리뷰 반영
  - roleId를 path variable로 받도록 변경
  - @PostMapping 을 @DeleteMapping으로 변경
  - controller단에서 ResponseDto를 리턴

- 출연자가 업체에게 지원 요청하는 관계 엔티티 관련 기획서 요구사항 구현 완료
  - 출연자
    - 출연자가 역할에 지원
    - 출연자가 지원한 역할들 확인
    - 지원 상태 별로 지원한 역할들 확인
    - 지원 취소
  - 업체
    - 특정 역할에 지원한 출연자들 확인
    - 업체 측에서 지원 승인, 거절
    - 출연 확정된 출연자들 확인
  
- API 테스트 완료

- 업체가 출연자에 출연 요청하는 기능은 기획이 명확해지면 구현할 예정

hyeneung committed:
refactor: ApplicationRequestCompany와 ApplicationRequestMember의 공통 부분을 ApplicationRequest 추상 클래스로 묶음

hyeneung committed:
style: 기존에 2개 있었던 BaseEntity를 하나로 만듦

createdAt과 modifiedAt은 DB의 인덱스로 쓰일 가능성 높음. 인덱스 성능 높이기 위해 not null constraint를 추가함. 추상 클래스로 명시함.

hyeneung committed:
feature: string을 enum으로 역직렬화하는 로직 커스터마이즈 및 적용

json에서 enum class에 정의되지 않은 문자열을 입력 받을 경우 null로 변환함으로써 null 여부로 유효성 검사 가능하도록 함

hyeneung committed:
feature: Controller Request DTO 작성

hyeneung committed:
feature: Service Request DTO 작성

Controller Request DTO에서 not null 체크를 했기에 따로 제약조건을 주진 않음.

hyeneung committed:
feature: DTO mapper 작성

hyeneung committed:
feature: ApplicationRequest 엔티티에서 발생 가능한 exception 정의 및 CustomException 작성

hyeneung committed:
feature: 출연 지원, 요청 관련 repository 작성

앱 무한스크롤을 위해 page 대신 slice 적용함

hyeneung committed:
feature: 출연 지원, 요청 관련 메서드 작성

hyeneung committed:
feature: 출연 지원 관련 로직 작성

출연자 중심 로직만 작성함. 업체에서 출연자에게 출연 요청하는 기능은 관련 기획이 명확해지면 추후에 구현할 예정.

hyeneung committed:
feature: 출연 지원 관련 controller 작성

업체측으로부터 출연자에게 출연 요청하는 것과 관련된 기능들은 추후에 해당 기획이 명확해지면 구현할 예정.

hyeneung committed:
style: roleId를 request body로 받기 위해 작성한 dto들 삭제

hyeneung committed:
feature: response dto 작성

hyeneung committed:
style: API 수정 하여 path variable 이용

hyeneung committed:
style: 불필요한 exception들 제거

기존에는 front측에서 업데이트 되지 않은 정보를 바탕으로 잘못된 요청을 내릴 경우를 상정하였으나, 사용자가 이미 지원한 역할에 대해 다시 지원할 일은 없다고 판단하여 제거함

hyeneung committed:
feature: enum 객체를 controller 단에서 validate

기존에는 enum을 request body로 받아 역직렬화 과정에서 validation이 가능했으나 path variable로 받기로 변경되어 controller 단에서 validate함

hyeneung committed:
bug: mapper의 source target 명시

test 과정에서 mapper의 문제로 null만 받았기에 source target을 명시하여 제대로 매핑되도록 함.

hyeneung committed:
feature: 특정 역할에 지원한 사용자들 확인한 결과 dto 작성

hyeneung committed:
feature: 역할과 연관된 출연자들 확인 관련 dto mapper 작성

hyeneung committed:
feature: 엔티티 내에 수정 로직 구현

사용자가 특정 역할에 지원하고 삭제할 때마다 role객체의 currentPersonnel 값 수정. 사용자의 요청에 업체에서 승인,거절할 때 해당 상태값 수정.

hyeneung committed:
feature: 업체 측에서 지원 승인, 거절하는 기능 관련 exception 작성

hyeneung committed:
feature: 특정 역할에 지원한 사용자들 조회, 특정 역할에 지원한 사용자의 지원 상태 변경 메서드 추가

hyeneung committed:
feature: 특정 역할에 지원한 사용자들 조회 하는 repository 메서드 작성

hyeneung committed:
feature: 특정 역할에 지원한 사용자들 조회, 특정 역할에 지원한 사용자의 지원 상태 변경 기능 구현

hyeneung committed:
feature: 특정 역할에 지원한 사용자들 조회, 특정 역할에 지원한 사용자의 지원 상태 변경 Controller 작성

hyeneung committed:
refactor: 엔티티 만들 필요 없는 것들 id 기반 쿼리로 대체

hyeneung committed:
feature: 출연 확정된 출연자들 보여주는 기능 위한 메서드 추가 #17

hyeneung committed:
feature: 출연 확정된 출연자들 보여주는 기능 위한 기능 추가 #17

hyeneung committed:
feature: 출연 확정된 출연자들 보여주는 기능 관련 controller 추가 및 null 처리 #17

hyeneung committed:
Merge branch 'develop' of https://github.com/pangseon/Extra into feature/attendance-api/attendance-management

hyeneung committed:
bug: mapstruct에서 필드를 인식하지 못하는 오류 수정

mapstruct와 lombok을 같이 쓰는 경우 annotationprocessor 선언 위치를 lombok이 앞에 가게끔 해야 한다고 함. 이대로 고쳤더니 필드 인식 잘됨.
https://stackoverflow.com/questions/63034956/mapstruct-no-property-named-packaging-exists-in-source-parameters

hyeneung committed:
feature: Put 요청 시에 applyStatus를 request body로 받기 위한 dto 작성

hyeneung committed:
feature: put 요청에 대한 dto mapper 작성

hyeneung committed:
feature: request body로 dto 받도록 전체 코드에 반영

hyeneung committed:
style: Member entity의 birtyday 필드명이 변경됨에 따라 responseDto도 변경

hyeneung committed:
refactor: 읽기 연산에 대해 read only임을 명시

shin0112 commented on src/main/java/com/example/extra/domain/applicationrequest/service/ApplicationRequestMemberService.java:
```diff
@@ -0,0 +1,36 @@
+        final long roleId,
+        Pageable pageable
+    );
+
+    void updateStatus(
+        Long roleId,
+        Long memberId,
+        ApplicationRequestMemberUpdateServiceRequestDto applicationRequestMemberUpdateServiceRequestDto
+    );
+
+    void deleteApplicationRequestMember(Long roleId);
```
final 붙이기..?

shin0112 commented on src/main/java/com/example/extra/domain/applicationrequest/service/impl/ApplicationRequestMemberServiceImpl.java:
```diff
@@ -0,0 +1,180 @@
+    @Transactional(readOnly = true)
+    public List<ApplicationRequestMemberReadServiceResponseDto> getAppliedRoles(Pageable pageable) {
+        Long memberId = 1L;
+        Slice<ApplicationRequestMember> applicationRequestMemberSlice =
+            applicationRequestMemberRepository.findAllByMemberId(
+                memberId,
+                pageable
+            );
```
조인문제로 인해서 컨벤션 때 외래키로 받는 경우 id가 아니라 객체 그 자체로 받기로 했습니다~

https://www.notion.so/Code-Convention-d54258b89dc0448ebbb5c785baf12023?pvs=4#5694b9413a954c168014c3ac8cf86a70

hyeneung replied:
제가 미처 확인을 못했네요 감사합니다 덕분에 공부도 많이 됐습니다!

hyeneung committed:
style: read only 파라미터에 대해 final 키워드 명시

hyeneung committed:
refactor: 외래키 기반 쿼리 대신 엔티티 기반 쿼리로 변경, id기반으로 엔티티 질의하는 공통된 부분 함수로 묶음

select from where id=? 이면 충분한데도 불필요한 join 연산 행해지는 것 예방하는 차원의 code convention에 따름.

hyeneung committed:
feature: 해당 엔티티의 기본키 전달하도록 response dto 재작성

application request member 테이블의 read 결과로 해당 테이블의 기본키 전달하고 추후 관련 기능 요청 시 해당 테이블의 기본키 기반으로 동작

hyeneung committed:
feature: response dto와 엔티티 mapper 작성

전달할 정보가 여러 테이블의 정보 필요로 하기에 mapper interface 내에 default method로 직접 mapper 정의

hyeneung committed:
feature: 변경된 response dto 사용하는 service 작성

hyeneung committed:
feature: 변경된 response dto 사용하는 controller 작성

hyeneung committed:
feature: 마감된 공고에 대해 지원시 exception 발생

shin0112 reviewed (approved):
수고가 많아용 👍
